### PR TITLE
Implement possibility of download url override using environment variables

### PR DESCRIPTION
**Pull request by baizulin**
Hi!

I've added possibility for the end user to upload binaries to custom location and serve them from there. This is extremely useful in certain cases, e.g. when github dl/ul is blocked (like at my workplace).

Kind regards,
Dennis

### CODE REVIEW

**Comment by juliangruber** [comment]
I'm curious, can you tell me why GitHub is blocked at your workplace?

File: README.md [hunk 1]
```diff
@@ -38,6 +38,15 @@ When `prebuild-install` is run via an `npm` script, options
 `--build-from-source`, `--debug` and `--download`, may be passed through via
 arguments given to the `npm` command.
 
+### Custom binaries
+The end user can override binary download location through environment variables in their .npmrc file.  
+The variable needs to meet the mask `% your package name %_binary_path`.
```
Comment (juliangruber) [inline README.md]: should be `binary_site`

File: README.md [hunk 2]
```diff
@@ -38,6 +38,15 @@ When `prebuild-install` is run via an `npm` script, options
...
 arguments given to the `npm` command.
 
+### Custom binaries
+The end user can override binary download location through environment variables in their .npmrc file.  
+The variable needs to meet the mask `% your package name %_binary_path`.  
+E.g.
+```
+leveldown_binary_site=http://overriden-host.com/overriden-path
+```
+Note that the package version subpath and file name will still be appended.
```
Comment (juliangruber) [inline README.md]: Can you give an example for the resulting url, for others?

**Comment by baizulin** [comment]
Copyright, Intellectual Property and sensitive data protection policies

**Comment by juliangruber** [comment]
Gotcha. Yes I think this addition would be great to have.

Can we maybe rename the property to make it clearer? What about

```
leveldown_prebuild_address=<HOST>/<PATH>
```

File: README.md (hunk 2 above)
Comment (baizulin) [reply README.md]: sure, updated readme

File: README.md (hunk 1 above)
Comment (baizulin) [reply README.md]: ah, right

**Comment by baizulin** [comment]
actually i've selected the prop name in accordance with similar one at https://github.com/sass/node-sass figured it'd be better this way

**Comment by baizulin** [comment]
@juliangruber dude?

**Comment by juliangruber** [comment]
does someone else have an opinion on this? i would rather prefer `NAME_prebuild_address`

**Comment by oohnoitz** [comment]
Not sure if you want to also consider compatibility with `node-pre-gyp` for this, but it supports the following:
- `npm install package --modulename_binary_host_mirror=https://...` 
- and `npm_config_modulename_binary_host_mirror` as an environment variable

Where `modulename` would be what was defined in the module's `package.json`.

**Comment by ralphtheninja** [comment]
> does someone else have an opinion on this? i would rather prefer NAME_prebuild_address

I don't have a strong opinion (naming is always hard though). But since it's actually a url, why not name it accordingly? My 2 cents are `NAME_prebuild_url`

**Comment by baizulin** [comment]
finally we have other opinions.
i'll just underline my point - as far as there is no common standard for the feature we are trying to implement i think we should go with the most popular/commonly used one. and i found it to be mostly used in the node-sass package (it's the industry standard for one of the most popular tasks, right?), the `sass_binary_site` property to be exact. also it has to be noted that we are not talking about a single specific precompiled binary but rather a custom location for binaries of different package versions for different environments, that's why i kept the version subpath and package name generation intact.
anyways i'm open for a discussion and interested in the changes getting to the master asap, so we need to come to the agreement on the prop name.

**Comment by oohnoitz** [comment]
I agree, it is annoying that there isn't any common standard for the naming convention. I only raised the question concerning compatibility with [node-pre-gyp](https://github.com/mapbox/node-pre-gyp) since:

1) it is the other tool that is used in some projects for building/installing native node addons, unless they roll their own like node-sass
2) it is already compatible with `node-pre-gyp` as it also uses the same `binary` structure inside of `package.json` for downloading the binaries
3) the `node-pre-gyp` [implementation](https://github.com/mapbox/node-pre-gyp/commit/2e5590ada88d71ec46a51493a9931f8e95cf8b1a) works almost the same way as you've implemented it with some slight difference since it uses the `binary` key in `package.json` and only overrides `binary.host`.

As a side note, noticed that the PR only applies this override if `binary` isn't set.

**Comment by juliangruber** [comment]
I'm for using the node-pre-gyp standard as prebuild-install is more related to node-pre-gyp than node-sass

**Comment by vweevers** [comment]
Using `npm_config_<module>_binary_host_mirror` makes perfect sense, because `prebuild-install` also supports the same `package.json` properties as `node-pre-gyp`.

The `*_binary_host_mirror` environment variable is not meant to override the URL structure, it's a mirror after all. So I'm -1 on `url`, `address` or `site` or anything that could be misconstrued as a fully qualified URL.

For symmetry with the `package.json` property though (`binary.host`), we could consider supporting both `*_binary_host` and `*_binary_host_mirror`.

**Comment by baizulin** [comment]
@juliangruber dude?

File: README.md [hunk 3]
```diff
@@ -38,6 +38,18 @@ When `prebuild-install` is run via an `npm` script, options
 `--build-from-source`, `--debug` and `--download`, may be passed through via
 arguments given to the `npm` command.
 
+### Custom binaries
+The end user can override binary download location through environment variables in their .npmrc file.  
+The variable needs to meet the mask `% your package name %_binary_site`. For example:
```
Comment (mathiask88) [inline README.md]: env name wrong here and below

Comment (baizulin) [reply README.md]: oh shoot, i knew i forgot something! thanks for noticing

**Comment by ralphtheninja** [comment]
I'd like to merge this. New minor I take it?

**Comment by mathiask88** [comment]
Just want to add, that this PR does not support nodejs 0.x due to `Object.assign()` so a merge would drop support for these two which are still supported by the current version AFAIR

**Comment by ralphtheninja** [comment]
> Just want to add, that this PR does not support nodejs 0.x due to Object.assign() so a merge would drop support for these two which are still supported by the current version AFAIR

We could use `xtend` instead, it's already a dependency.

**Comment by mathiask88** [comment]
LGTM @ralphtheninja

File: test/util-test.js [hunk 4]
```diff
@@ -75,9 +76,34 @@ test('urlTemplate() returns different templates based on pkg and rc', function (
   }
   var t6 = util.urlTemplate(o6)
   t.equal(t6, 'http://foo.com/w00t/{name}-{major}.{minor}-{runtime}-v{abi}-{platform}-{arch}.tar.gz', 'pkg.binary.package_name is added after host and remote_path, custom format')
-  var o7 = {pkg: require('../package.json'), download: true}
+  var o7 = {
+    pkg: xtend({}, require('../package.json')),
```
Comment (vweevers) [inline test/util-test.js]: The first argument `{}` is not necessary

Comment (juliangruber) [reply test/util-test.js]: true, it's different from `Object.assign` in that respect

Comment (ralphtheninja) [reply test/util-test.js]: @vweevers Removed in commit after merge.

**Pull request merged**